feat: code commenting #9369

ravikp7 opened the pull request:
## Description
Adds code commenting in JS objects code editor and JS fields. Users can use `Cmd + /` on Mac and `Ctrl + /` on other systems to comment/uncomment code now.

Fixes #9369



## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- Manual
- Jest

### Test Plan
- [ ] https://github.com/appsmithorg/TestSmith/issues/2120
- [ ] https://github.com/appsmithorg/TestSmith/issues/2121
- [ ] https://github.com/appsmithorg/TestSmith/issues/2122

### Issues raised during DP testing
- [ ] https://github.com/appsmithorg/appsmith/pull/18667#issuecomment-1348354145


## Checklist:
### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag


### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+
+const JS_FIELD_BEGIN = "{{";
+
+const nonWS = /[^\s\u00a0]/;
```
can we have a more descriptive name for this variable?

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+/**
+ * Gives index of the first non whitespace character in the line
+ **/
+function firstNonWS(str: string) {
```
can we have a more descriptive name for this function?

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+ **/
+function firstNonWS(str: string) {
+  const found = str.search(nonWS);
+  return found == -1 ? 0 : found;
```
can we have strict equality here?

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+function probablyInsideString(
+  cm: CodeMirror.Editor,
+  pos: CodeMirror.Position,
+  line: any,
```
can we add a type for this?

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+}
+
+// Rough heuristic to try and detect lines that are part of multi-line string
+function probablyInsideString(
```
can we give a better name for the function?

ravikp7 replied:
Can't come up with a better name, also not completely sure what it does, copied from the library. Let's leave this?

rimildeyjsr replied:
Can we assess the inputs and outputs to understand what it does? Else we can skip this

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+  );
+}
+
+function getMode(cm: CodeMirror.Editor, pos: CodeMirror.Position) {
```
this is a more generic function specific to the editor, can we move this to `codeEditorutils` file?

ravikp7 replied:
It's copied from the library, so kept it here only.

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+
+function getMode(cm: CodeMirror.Editor, pos: CodeMirror.Position) {
+  const mode = cm.getMode();
+  return mode.useInnerComments === false || !mode.innerMode
```
nitpick - we can say `!mode.useInnerComments` here

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+    const self = this,
+      mode = getMode(self, from);
+    const firstLine = self.getLine(from.line);
+    if (firstLine == null || probablyInsideString(self, from, firstLine))
```
nitpick - we can say `!firstLine` here

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+    const end = Math.min(
+      to.ch != 0 || to.line == from.line ? to.line + 1 : to.line,
+      self.lastLine() + 1,
+    );
```
can we move this part to its own function, getEndLinePosition?
also can we rename the variable to something more descriptive, this is a little hard to understand

ravikp7 replied:
Moved to a function

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+      to.ch != 0 || to.line == from.line ? to.line + 1 : to.line,
+      self.lastLine() + 1,
+    );
+    const pad = options.padding == null ? " " : options.padding;
```
nitpick - this can be `options.padding || " "`

also can we rename this constant to padding ?

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+      self.lastLine() + 1,
+    );
+    const pad = options.padding == null ? " " : options.padding;
+    const blankLines = options.commentBlankLines || from.line == to.line;
```
lets use strict equality here

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+        //     baseString = whitespace;
+        //   }
+        // }
+        /* Changes end */
```
do we need these changes? else lets get rid of them

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+          /* Added below code to handle baseString individually for each line */
+          const baseString =
+            line.search(nonWS) === -1 ? line : line.slice(0, firstNonWS(line));
+          /* Changes end */
```
i think one comment line is enough, we don't need the changes end comment

ravikp7 replied:
I meant my changes end here. Like, this block has been inserted in the copied code.

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+            line.search(nonWS) === -1 ? line : line.slice(0, firstNonWS(line));
+          /* Changes end */
+
+          const cut = (baseString || "").length;
```
lets have a more descriptive constant here

ravikp7 replied:
Renamed with `offset`

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+          if (!blankLines && !nonWS.test(line)) continue;
+          /* Commented below line to handle baseString individually for each line */
+          // if (line.slice(0, cut) != baseString) cut = firstNonWS(line);
+          /* Changes end */
```
here also, we can get rid of the changes end comment

rimildeyjsr replied:
also if we don't need this, lets get rid of this

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+        for (let i = from.line; i < end; ++i) {
+          const line = self.getLine(i);
+
+          /* Added below code to handle baseString individually for each line */
```
lets make this a line comment

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+          // if (line.slice(0, cut) != baseString) cut = firstNonWS(line);
+          /* Changes end */
+
+          /* Handle JS field lines starting with {{ */
```
lets change this to a line comment

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+            );
+            continue;
+          }
+          /* End changes for JS field handling */
```
here also, we can get rid of the changes end comment

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+        for (let i = from.line; i < end; ++i) {
+          const line = self.getLine(i);
+          if (blankLines || nonWS.test(line)) {
+            /* Handle JS field lines starting with {{ */
```
lets make this a line comment

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+              );
+              continue;
+            }
+            /* End changes for JS field handling */
```
here also, we can get rid of the changes end comment

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+    const end = Math.min(
+      to.ch != 0 || to.line == from.line ? to.line : to.line - 1,
+      self.lastLine(),
+    );
```
lets move this to a function

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+    const lineString = options.lineComment || mode.lineComment;
+    const lines: string[] = [];
+    const pad = options.padding == null ? " " : options.padding;
+    let didSomething;
```
lets use a more descriptive variable name

ravikp7 replied:
Renamed with `didCommentCode`

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+    // Try finding line comments
+    const lineString = options.lineComment || mode.lineComment;
+    const lines: string[] = [];
+    const pad = options.padding == null ? " " : options.padding;
```
can we rename this to `options.padding || " "`

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+        //   !/comment/.test(self.getTokenTypeAt(CodeMirror.Pos(i, found + 1)))
+        // )
+        //   found = -1;
+        /* Change ends */
```
lets get rid of this if we don't need it

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+    const lead = options.blockCommentLead || mode.blockCommentLead;
+    const startLine = self.getLine(start);
+    const open = startLine.indexOf(startString);
+    if (open == -1) return false;
```
lets use strict equality here

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+import { isMacOrIOS } from "utils/helpers";
+
+export const getCodeCommentKeyMap = () => {
+  return isMacOrIOS() ? "Cmd-/" : "Ctrl-/";
```
you can have a object lookup that returns the shortcut needed

ravikp7 replied:
I think it's same for linux and windows, so went with `isMac` condition. We shouldn't need to handle linux and windows separately, right?

rimildeyjsr replied:
Its better we handle for all cases

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+    });
+  };
+
+  cm.uncomment = function(from, to, options) {
```
can we move this part to its own function

ravikp7 replied:
same comment as above.

rimildeyjsr replied:
replied above

rimildeyjsr commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComments.test.ts:
```diff
@@ -0,0 +1,180 @@
+import CodeMirror from "codemirror";
+import "codemirror/addon/comment/comment";
+import { handleCodeComment } from "./codeComment";
```
lets add cypress test cases for this as well, and get them reviewed by the QA team

rimildeyjsr replied:
end to end testing will help us test the shortcut key scenarios, I think it would be helpful to have the E2E tests

ravikp7 replied:
Pushed E2E test

rimildeyjsr commented:
@ravikp7 - you will have to add the title of the PR as  `feat:<title>` for the semantic PR check to pass.

Also, can we add a detailed PR description, with the shortcut keys described? This helps @theparthacus while preparing our release notes and serves as a future reference for anyone who needs to come back to this PR

ravikp7 commented:
@rimildeyjsr Most of the comments on the PR are on the code that I’ve copied from the codemirror comment library, and I know that code is pretty badly written. I just didn’t want to do any modifications to their code as it may introduce some unnecessary bugs. I want to keep it as close to the original code as possible, so I’ve put explicit comments where I’ve added my code or commented out their code. What do you think?

ravikp7 commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+export const handleCodeComment = (cm: CodeMirror.Editor) => {
+  const noOptions: CodeMirror.CommentOptions = {};
+
+  cm.lineComment = function(from, to, options = noOptions) {
```
We're overriding the existing `lineComment` function here. Not sure, what you're suggesting.

rimildeyjsr replied:
so we can have something like - 
const performLineCommenting = () => {}

and then set this function to cm.lineComment. We can do the same for uncomment as well

In my opinion, it makes the code little but readable, but let me know if you feel otherwise!

ravikp7 commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+    const startString = options.blockCommentStart || mode.blockCommentStart;
+    const endString = options.blockCommentEnd || mode.blockCommentEnd;
+    if (!startString || !endString) return false;
+    const lead = options.blockCommentLead || mode.blockCommentLead;
```
Renamed with `blockCommentLead`

ravikp7 commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+      .slice(to.ch)
+      .lastIndexOf(startString, firstEnd - to.ch);
+    lastStart =
+      firstEnd == -1 || almostLastStart == -1 ? -1 : to.ch + almostLastStart;
```
It's very difficult to understand and we're not using block commenting, so I'm not in favor of modifying this code.

ravikp7 commented on app/client/src/components/editorComponents/CodeEditor/utils/codeComment.ts:
```diff
@@ -0,0 +1,291 @@
+          );
+        }
+    });
+    return true;
```
I'm not sure of that, returning `false` also, doesn't affect anything, so just leaving it as it was.

ravikp7 commented:
@AnandiKulkarni Can you please help with E2E testing for this?

ravikp7 commented:
/ok-to-test sha=2e80158e28d87f23d43cfbe6e76a0c82859557c2

Aishwarya-U-R reviewed (commented):
Cypress looks good!

ravikp7 commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/VisualTests/JSEditorComment_spec.js:
```diff
@@ -0,0 +1,50 @@
+  agHelper = ObjectsRegistry.AggregateHelper;
+
+describe("JSEditor Comment - Visual tests", () => {
+  it("comments code on the editor", () => {
```
added

ramsaptami commented:
- [ ] Autocomplete does not work on a Windows machine. Selecting anything from the dropdown, closes the pop-up but does not insert anything

ravikp7 commented:
/ok-to-test sha=07cbc33f99063bde71deefe54c3fa5ffbe6dca56

ravikp7 commented:
/ok-to-test sha=0090c53228675f1d82c1bfe0f84044346e78db65

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Autocomplete/PropertyPaneSuggestion_spec.js:
```diff
@@ -0,0 +1,32 @@
+    cy.addDsl(dsl);
+  });
+
+  it("1. Should show Property Pane Suggestions on / command", () => {
```
Please consider writing tests in typescript. There is a /BugTests/Autocomplete_spec.ts which can be used to refer to the ts helpers available for validating this.

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Autocomplete/PropertyPaneSuggestion_spec.ts:
```diff
@@ -0,0 +1,35 @@
+    EntityExplorer.SelectEntityByName("Button1", "Widgets");
+    PropertyPane.TypeTextIntoField("Label", "{{appsmith}}");
+    PropertyPane.TypeTextIntoField("Label", "/");
+    cy.get("ul.CodeMirror-hints")
+      .contains("New Binding")
+      .should("exist")
+      .click();
```
Can we use below instead:
`    agHelper.GetNAssertElementText(locator._hints, "New Binding");
`
`    agHelper.GetNClickByContains(locator._hints, "New Binding);
`

ravikp7 replied:
Done, thanks!

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Autocomplete/PropertyPaneSuggestion_spec.ts:
```diff
@@ -0,0 +1,35 @@
+    EntityExplorer.SelectEntityByName("Button1", "Widgets");
+    PropertyPane.TypeTextIntoField("Label", "{{appsmith}}");
+    PropertyPane.TypeTextIntoField("Label", "{{");
+    cy.get("ul.CodeMirror-hints")
+      .contains("appsmith")
+      .should("exist")
+      .click();
```
Same as above!

ravikp7 replied:
Done

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Autocomplete/PropertyPaneSuggestion_spec.ts:
```diff
@@ -0,0 +1,35 @@
+
+    PropertyPane.ValidatePropertyFieldValue("Label", "{{appsmith}}");
+  });
+});
```
Can you please also move the other related Autocomplete specs (js/ts) from BugTests into this Autocomplete folder please?

ravikp7 replied:
Done

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/CodeComment/PropertyPaneCodeComment_spec.ts:
```diff
@@ -0,0 +1,28 @@
+
+  it("1. Should comment code in Property Pane", () => {
+    EntityExplorer.SelectEntityByName("Button1", "Widgets");
+    PropertyPane.TypeTextIntoField("Label", "{{appsmith}}");
```
TypeTextIntoField usually doesnt require - closing '}}', Hope these cases are passing!

ravikp7 replied:
I've added `{{ }}` on purpose to test code commenting with JS snippets.

Aishwarya-U-R commented on app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Autocomplete/PropertyPaneSuggestion_spec.ts:
```diff
@@ -0,0 +1,40 @@
+    });
+  });
+
+  it("1. Should show Property Pane Suggestions on / command", () => {
```
Is this feature automated against bug or test feature? If bug - can you please add the bug id to `it` or if feature - the related test smith id to `it` please?
So tomorrow if we need to automate some bug/test case when it comes to Automation team - we know its already covered & do not duplicate the work

Rishabh-Rathod reviewed (commented):
Unit tests LGTM

AnandiKulkarni commented:
@ravikp7 below commenting in query is wrong 
Currently it is working as - select * from us
ers where name = {{ --input2.text,
input.text}}
ex: -- select * from users where name = {{ //input2.text,
input.text}}
<img width="1440" alt="Screenshot 2022-12-22 at 10 57 50 PM" src="https://user-images.githubusercontent.com/80756091/209195253-7f5db4c3-6abe-44a2-aba5-b714f00bc254.png">

AnandiKulkarni commented:
@ravikp7 Please find below recording on list widget cmd+l works outside {{}}, which does not evaluate anything but it works only before label and value and not before children, behaviour is very abrupt 

https://user-images.githubusercontent.com/80756091/209196871-f295647d-6a15-4e4d-b23c-d90dd09b1d83.mov

ravikp7 commented:
> @ravikp7 below commenting in query is wrong Currently it is working as - select * from us ers where name = {{ --input2.text, input.text}} ex: -- select * from users where name = {{ //input2.text, input.text}} <img alt="Screenshot 2022-12-22 at 10 57 50 PM" width="1440" src="https://user-images.githubusercontent.com/80756091/209195253-7f5db4c3-6abe-44a2-aba5-b714f00bc254.png">

@AnandiKulkarni This happened only when the cursor was placed inside the JS binding. Thanks for catching it. I've pushed a fix for it and added another unit test for this case.

ravikp7 commented:
> @ravikp7 Please find below recording on list widget cmd+l works outside {{}}, which does not evaluate anything but it works only before label and value and not before children, behaviour is very abrupt
> 
>  Screen.Recording.2022-12-22.at.11.20.18.PM.mov

@AnandiKulkarni I've fixed this issue as well and added unit tests, only the lines with text binding will get commented out with comments inside `{{ }}`. And, for the rest of the lines commenting will not work where `{{ }}` is not used because those lines are not JS, but just plain text.

ravikp7 commented:
/ok-to-test sha=e0110bb8f6f1f29f05c50b12ec4d3fd1f7decb16

ravikp7 commented:
/ok-to-test sha=564de767b24e48424fe5296b642dbeca0e03cb8d

ravikp7 commented:
/perf-test sha=564de767b24e48424fe5296b642dbeca0e03cb8d

ravikp7 commented:
/ok-to-test sha=aeba6b22199408be4bc3473de29b484179517d94

ravikp7 commented:
/ok-to-test sha=4153fe29ea934db0d49bb0d4b8736170876d0331

ravikp7 commented:
/ok-to-test sha=004087ad10f4e66dee628a97b33bc5d7a1d3dfd7

ravikp7 commented:
/ok-to-test sha=004087ad10f4e66dee628a97b33bc5d7a1d3dfd7

ravikp7 commented:
/ok-to-test sha=994350755eb7281653062fed5dc6eed00aabc923

ravikp7 commented:
/ok-to-test sha=943ebfd60b255f8be0ec1102656ab7258c516af6

ravikp7 commented:
/ok-to-test sha=6b6b26eed7d2342beecb724080e2eb2c3d716bef
